manifests/agent/service.pp

```diff
@@ -3,58 +3,32 @@
 
   case $::puppet::runmode {
     'service': {
-      service {'puppet':
-        ensure     => running,
-        name       => $puppet::service_name,
-        hasstatus  => true,
-        hasrestart => $puppet::agent_restart_command != undef,
-        enable     => true,
-        restart    => $puppet::agent_restart_command,
-      }
-
-      if $::osfamily != 'windows' {
-        cron { 'puppet':
-          ensure => absent,
-        }
+      class { '::puppet::agent::service::daemon':  enabled => true  }
+      class { '::puppet::agent::service::cron':    enabled => false }
+      if $::kernel == 'Linux' {
+        class { '::puppet::agent::service::systemd': enabled => false }
       }
     }
     'cron': {
-      service {'puppet':
-        ensure    => stopped,
-        name      => $puppet::service_name,
-        hasstatus => true,
-        enable    => false,
-      }
-
-      $command = $puppet::cron_cmd ? {
-        undef   => "/usr/bin/env puppet agent --config ${puppet::dir}/puppet.conf --onetime --no-daemonize",
-        default => $puppet::cron_cmd,
+      class { '::puppet::agent::service::daemon':  enabled => false }
+      class { '::puppet::agent::service::cron':    enabled => true  }
+      if $::kernel == 'Linux' {
+        class { '::puppet::agent::service::systemd': enabled => false }
       }
-
-      if $::osfamily == 'windows' {
-        fail("Currently we don't support setting cron on windows.")
-      } else {
-        $times = ip_to_cron($puppet::runinterval)
-        cron { 'puppet':
-          command => $command,
-          user    => root,
-          hour    => $times[0],
-          minute  => $times[1],
-        }
+    }
+    'systemd.timer', 'systemd': {
+      if $::kernel != 'Linux' {
+        fail("Runmode of ${puppet::runmode} not supported on ${::kernel} operating systems!")
       }
+      class { '::puppet::agent::service::daemon':  enabled => false }
+      class { '::puppet::agent::service::cron':    enabled => false }
```

Comment:
This will fail on Archlinux - there is no cron provider for puppet to use. Can we make it optional? $disable_modes = true ?